pkg/controller/messaginginfra/broker/controller.go

```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright 2020, EnMasse authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+
+package broker
+
+import (
+	"context"
+	"fmt"
+
+	logr "github.com/go-logr/logr"
+
+	v1beta2 "github.com/enmasseproject/enmasse/pkg/apis/enmasse/v1beta2"
+	"github.com/enmasseproject/enmasse/pkg/controller/messaginginfra/cert"
+	"github.com/enmasseproject/enmasse/pkg/controller/messaginginfra/common"
+	"github.com/enmasseproject/enmasse/pkg/util"
+	"github.com/enmasseproject/enmasse/pkg/util/install"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	resource "k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type BrokerController struct {
+	client         client.Client
+	scheme         *runtime.Scheme
+	certController *cert.CertController
+}
+
+func NewBrokerController(client client.Client, scheme *runtime.Scheme, certController *cert.CertController) *BrokerController {
+	return &BrokerController{
+		client:         client,
+		scheme:         scheme,
+		certController: certController,
+	}
+}
+
+func getBrokerLabels(infra *v1beta2.MessagingInfra) map[string]string {
+	labels := make(map[string]string, 0)
+	labels["component"] = fmt.Sprintf("broker-%s", infra.Name)
+	labels["app"] = "enmasse"
+
+	return labels
+}
+
+/*
+ * Reconciles the broker instances for an instance of shared infrastructure.
+ *
+ * Each instance of a broker is created as a statefulset. If we want to support HA in the future, each statefulset can use replicas to configure HA.
+ */
+func (b *BrokerController) ReconcileBrokers(ctx context.Context, logger logr.Logger, infra *v1beta2.MessagingInfra) error {
+	setDefaultBrokerScalingStrategy(&infra.Spec.Broker)
+	logger.Info("Reconciling brokers", "broker", infra.Spec.Broker)
+
+	labels := getBrokerLabels(infra)
+
+	// Update broker condition
+	brokersCreated := infra.Status.GetMessagingInfraCondition(v1beta2.MessagingInfraBrokersCreated)
+	brokersCreated.SetStatus(corev1.ConditionTrue, "", "")
+
+	return common.WithConditionUpdate(brokersCreated, func() error {
+		brokers := appsv1.StatefulSetList{}
+		err := b.client.List(ctx, &brokers, client.InNamespace(infra.Namespace), client.MatchingLabels(labels))
+		if err != nil {
+			return nil
+		}
+
+		for _, broker := range brokers.Items {
+			err := b.reconcileBroker(ctx, logger, infra, &broker)
+			if err != nil {
+				return err
+			}
+		}
+
+		toCreate := numBrokersToCreate(infra.Spec.Broker.ScalingStrategy, brokers.Items)
+		if toCreate > 0 {
+			logger.Info("Creating brokers", "toCreate", toCreate)
+			for i := 0; i < toCreate; i++ {
+				statefulset := &appsv1.StatefulSet{
+					ObjectMeta: metav1.ObjectMeta{Namespace: infra.Namespace, Name: fmt.Sprintf("broker-%s-%s", infra.Name, util.RandomBrokerName())},
```

Comment:
If creating a single stateful set would fail, you would re-queued and re-try the stateful set creation. Having a random component in the name, would creating X additional, rather than the remaining stateful sets.

Reply:
I'm not sure I understand the concern. I think the following three scenarios are working right now:

Scenario 1: Statefulset does not get created, reconcile returns error and reconcile is re-run with the initial listBrokers
Scenario 2: Statefulset gets created, but error in creating other resources lead to error. Reconcile returns error, but statefulset is re-read during listBrokers and reconcile continues
Scenario 3: Statefulset gets created, other resources get created. Things proceed as normal.

Reply:
Ah ok, I see now. Maybe adding a few comments would have helped understanding :)